options.go

```diff
@@ -0,0 +1,16 @@
+package net
+
+import (
+	"context"
+)
+
+type noDialCtxKey struct{}
+
+// NoDial is a context option that instructs the network to not attempt a new
+// dial when opening a stream. The value of the key should be a string indicating
+// the source of the option.
+var NoDial = noDialCtxKey{}
+
+func WithNoDial(ctx context.Context, src string) {
+	context.WithValue(ctx, NoDial, src)
+}
```

Comment:
So, personally, I'd not expose the `NoDial` var but instead expose a `GetNoDial(ctx) (nodial bool, why string)` function. That way, we can change how we implement all this internally.

For example, we may want to "lift" all of our options into an option struct if we end up chaining contexts too much (searching for a value in a context is O(n)).

Reply:
ok, will do.

Reply:
(I mean, this just comes down to "patterns" but I believe this is the more common pattern.)

Reply:
done.